### PR TITLE
Framework: Reuse computed attributes from `site` to `getSite` selector

### PR DESCRIPTION
**Pull request by ehg**
We don't want computed attributes in the state tree, so let's put them in the `getSite` selector.

- Extracted computed attrs logic into its own module
- Updated `site` to use new module
- Update `getSite` selector to append computed attrs
- Compose `getSite`'s site with the selectors we already have
- Remove computed attrs from schema

/cc @mtias 

TODO:
- [x] Fix/add tests

To test:
 - Since this affects anything that uses `site`, both in the old manner (`sites-list`), and anything using the `getSite` selector. Make sure the various sections work as expected.

Test live: https://calypso.live/?branch=add/computed-attrs-to-getSite-selector

### CODE REVIEW

File: client/lib/site/computed-attributes.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Internal dependencies
+ */
+import config from 'config';
+import { isHttps } from 'lib/url';
+
+export default function( site ) {
+	const attributes = {};
+	/**
+	 * If the user has no access to site.options create it as an empty
```
Comment (mtias) [inline client/lib/site/computed-attributes.js]: Let's use `//` comments while we are here.

**Comment by aduth** [comment]
Should we create independent selectors, like we've done with [`getSiteSlug`](https://github.com/Automattic/wp-calypso/blob/7e460ac398dce3a8e548f50d8d934ece45298165/client/state/sites/selectors.js#L79-L97), or at the very least offer the decorated object separate from the raw as we've done with [`getNormalizedPost`](https://github.com/Automattic/wp-calypso/blob/7e460ac398dce3a8e548f50d8d934ece45298165/client/state/posts/selectors.js#L37-L66)?

I've personally taken more of a [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) approach with reimplementing computed properties as selectors, as it might present some candidates for trimming and reevaluating our usage (e.g. in implementing slugs, I [followed-up](https://github.com/Automattic/wp-calypso/pull/4055#discussion_r56454735) on whether we needed conflict detection and simplified the logic some).

File: client/lib/site/computed-attributes.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		attributes.slug = attributes.options.unmapped_url.replace( /^https?:\/\//, '' );
+		attributes.domain = attributes.slug;
+	}
+
+	// The 'standard' post format is saved as an option of '0'
+	if ( ! attributes.options.default_post_format || attributes.options.default_post_format === '0' ) {
+		attributes.options.default_post_format = 'standard';
+	}
+	attributes.is_previewable = !! (
+		config.isEnabled( 'preview-layout' ) &&
```
Comment (mtias) [inline client/lib/site/computed-attributes.js]: Do we need this config check here? Seems a bit out of place.

File: client/lib/site/computed-attributes.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		attributes.options.unmapped_url &&
+		! site.is_vip &&
+		isHttps( attributes.options.unmapped_url )
```
Comment (mtias) [inline client/lib/site/computed-attributes.js]: Since this is quite ad hoc (and hopefully temporary) perhaps we should have an `isPreviewableSite` selector instead?

File: client/lib/site/computed-attributes.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	);
+	attributes.is_customizable = !! (
```
Comment (mtias) [inline client/lib/site/computed-attributes.js]: I think we should remove this one and use `currentUserCan` as much as possible in the code.

File: client/state/sites/selectors.js [hunk 5]
```diff
@@ -23,7 +24,8 @@ import createSelector from 'lib/create-selector';
  * @return {?Object}        Site object
  */
 export function getSite( state, siteId ) {
-	return state.sites.items[ siteId ] || null;
+	const site = state.sites.items[ siteId ];
+	return site ? Object.assign( site, getAttributes( site ) ) : null;
```
Comment (mtias) [inline client/state/sites/selectors.js]: Should we memoize this selector to avoid computing things every time we getSites (which is a lot)?

**Comment by mtias** [comment]
@aduth I think in this case the value of a selector that returns the raw data is much less than for posts (where raw is crucial in some contexts). I don't see anyone consuming the site object without these normalized attributes. I will also be open to renaming the attributes to camelCase and ditching anything we don't need (this last piece could hopefully transition into `fields` arguments so that the API doesn't even generate them).

File: client/lib/site/computed-attributes.js (hunk 3 above)
Comment (ehg) [inline client/lib/site/computed-attributes.js]: Yeah, since pulling that out will require `connect`ing the my-sites sidebar, I'd rather leave it until another PR.

Comment (mtias) [inline client/lib/site/computed-attributes.js]: Right, let's add a comment then, otherwise I fear we'll lose it :)

Comment (ehg) [inline client/lib/site/computed-attributes.js]: Yep ;) Already done that, but haven't pushed.

File: client/lib/site/computed-attributes.js (hunk 4 above)
Comment (ehg) [inline client/lib/site/computed-attributes.js]: Added a comment for this as well.

File: client/lib/site/computed-attributes.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+	// If a WordPress.com site has a mapped domain create a `wpcom_url`
+	// attribute to allow site selection with either domain.
+	if ( attributes.options && attributes.options.is_mapped_domain && ! site.is_jetpack ) {
+		attributes.wpcom_url = site.options.unmapped_url.replace( /^https?:\/\//, '' );
+	}
+
+	// If a site has an `is_redirect` property use the `unmapped_url`
+	// for the slug and domain to match the wordpress.com original site.
+	if ( ( attributes.options && attributes.options.is_redirect ) || site.hasConflict ) {
```
Comment (aduth) [inline client/lib/site/computed-attributes.js]: Where is `hasConflict` coming from in the Redux sites case?

Comment (ehg) [inline client/lib/site/computed-attributes.js]: It's being set in `sites-list`, in`initialize()` and `update()`, using https://github.com/Automattic/wp-calypso/blob/master/client/lib/sites-list/list.js#L156

Should we override attributes we already have selectors for in Redux? Maybe marking the code in `sites-list` and `site` as deprecated — do not modify.

**Comment by ehg** [comment]
> Should we create independent selectors, like we've done with getSiteSlug

I think offering a completely normalised `getSite` is going to ease the transition from `sites-list`. We can still move the old logic into separate selectors, and use them in `getSite`. We end up with some code duplication this way, but I believe it'll allow us to separately tackle some of the more onerous aspects of `sites-list`, e.g. polling — without having to worry about the `site` objects too much.

**Comment by aduth** [comment]
@ehg : I'd agree with the idea that a normalized site object would be easier to consume from and transition to for developers (evidenced by confusion around and reintroduction of these attributes in the first place 😄 ). Individual selectors have a number of advantages, but "obviousness" (ease of use) isn't really one of them. Composing attributes into the normalized object using individual selectors might be a compromise of these two options, but then there may be confusion around which is preferred (do we have a preference, if the developer has gone through the effort to distinguish them?).

Regardless, our normalized site object should behave exactly as it does in `sites-list` if this is the path we're to pursue, regardless of its source. We should pretend that we're not passing the already-decorated site object into the sites reducer, as eventually this will be the case when we're requesting sites data directly in Redux. With this in mind, and the dependency on `hasConflict`, there's the potential for future issues when finally migrating away from `sites-list`. The solution here may be, as you mention, to assign the `slug` attribute (and `domain`?) using the state selector. Otherwise, we'll want to reimplement `hasConflict` to accommodate the Redux site object.

File: client/lib/site/computed-attributes.js (hunk 6 above)
Comment (aduth) [inline client/lib/site/computed-attributes.js]: >It's being set in `sites-list`, in`initialize()` and `update()`

But this attribute will be (rightfully) [stripped from the state object](https://github.com/Automattic/wp-calypso/blob/8527d55850218ce2b8429cc6fe1331e3628c7a85/client/state/sites/reducer.js#L55) when received into the reducer.

File: client/lib/site/computed-attributes.js (hunk 3 above)
Comment (ockham) [inline client/lib/site/computed-attributes.js]: @ehg Keep me posted about `connect`ing the sidebar, I have a PR (#6165) that turns our showcase helpers into selectors, so that'd enable me to use a `getCustomizeUrl` selector there.

File: client/lib/site/computed-attributes.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+	//TODO:(ehg) Pull out into selector when my-sites/sidebar is connected
+	attributes.is_previewable = !! (
+		config.isEnabled( 'preview-layout' ) &&
+		attributes.options.unmapped_url &&
+		! site.is_vip &&
+		isHttps( attributes.options.unmapped_url )
+	);
+
+	//TODO:(ehg) Replace instances with canCurrentUser selector when my-sites/sidebar is connected
```
Comment (ockham) [inline client/lib/site/computed-attributes.js]: ❤️ 

Will also mean turning the `GuidedTours` config into something componentish and connecting it, no? https://github.com/Automattic/wp-calypso/pull/5665/files?diff=unified#diff-147ca0cbc1b586bbacaec2b9c9c6f8f9R46

File: client/state/sites/selectors.js [hunk 8]
```diff
@@ -145,7 +170,7 @@ export function isJetpackMinimumVersion( state, siteId, version ) {
  * @return {?String}        Site slug
  */
 export function getSiteSlug( state, siteId ) {
-	const site = getSite( state, siteId );
+	const site = getRawSite( state, siteId );
```
Comment (mtias) [inline client/state/sites/selectors.js]: Shouldn't `getSiteSlug` be using the computed value?

Comment (mtias) [inline client/state/sites/selectors.js]: The code below that checks site.options should be made redundant now and getSiteSLug should just return the `slug` attribute of `getSite`.

Comment (nylen) [inline client/state/sites/selectors.js]: `getSite` uses `getSiteSlug`, and it would be less performant to calculate all the other options and then throw them away.

Comment (mtias) [inline client/state/sites/selectors.js]: Right, but I mean that `getSiteSlug` should just do `return getSite( state, siteId ).slug;` with slug being computed by the new `getSite`.

Comment (ehg) [inline client/state/sites/selectors.js]: So from memory, I envisaged composing the selectors as the best solution, eventually getting rid of legacy computed attributes when we nuke `sites-list` — hence the redundancy for now.

File: client/state/sites/selectors.js [hunk 9]
```diff
@@ -20,17 +21,41 @@ import {
...
+		const site = state.sites.items[ siteId ];
+
+		if ( site ) {
+			Object.assign( site, getComputedAttributes( site ) );
+			site.hasConflict = isSiteConflicting( state, siteId );
+			site.slug = getSiteSlug( state, siteId );
+			site.domain = getSiteDomain( state, siteId );
+		}
+
+		return site || null;
```
Comment (aduth) [inline client/state/sites/selectors.js]: Let's just return early with null if the site is not set so we don't have the extra level of indentation in the condition above.

File: client/state/sites/selectors.js [hunk 10]
```diff
@@ -20,17 +21,41 @@ import {
...
  * @param  {Number}  siteId Site ID
  * @return {?Object}        Site object
  */
-export function getSite( state, siteId ) {
+export const getSite = createSelector(
+	( state, siteId ) => {
+		const site = state.sites.items[ siteId ];
+
+		if ( site ) {
+			Object.assign( site, getComputedAttributes( site ) );
```
Comment (aduth) [inline client/state/sites/selectors.js]: This is mutating state directly with the `Object.assign`.

File: client/state/sites/selectors.js [hunk 11]
```diff
@@ -20,17 +21,41 @@ import {
...
+			site.hasConflict = isSiteConflicting( state, siteId );
+			site.slug = getSiteSlug( state, siteId );
+			site.domain = getSiteDomain( state, siteId );
+		}
+
+		return site || null;
+	},
```
Comment (aduth) [inline client/state/sites/selectors.js]: With recommended changes:

```js
const site = state.sites.items[ siteId ];
if ( ! site ) {
	return null;
}

return {
	...site,
	...getComputedAttributes( site ),
	hasConflict: isSiteConflicting( state, siteId ),
	slug: getSiteSlug( state, siteId ),
	getSiteDomain( state, siteId )
};
```

File: client/lib/site/computed-attributes.js [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+	// If the user has no access to site.options create it as an empty
+	// attribute to avoid potential errors when trying to access its sub properties
+	attributes.options = site.options || {};
+
+	// Add URL without protocol as a `domain` attribute
+	if ( site.URL ) {
+		attributes.domain = site.URL.replace( /^https?:\/\//, '' );
+		attributes.slug = attributes.domain.replace( /\//g, '::' );
+	}
+	attributes.title = trim( site.name ) || attributes.domain;
```
Comment (aduth) [inline client/lib/site/computed-attributes.js]: Can `site.name` be anything other than a string? If not, we could just use the base [`String#trim`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/Trim).

File: client/lib/site/index.js [hunk 13]
```diff
@@ -86,46 +87,7 @@ Site.prototype.set = function( attributes ) {
  * Update any computed attributes
  */
 Site.prototype.updateComputedAttributes = function() {
-	/**
```
Comment (aduth) [inline client/lib/site/index.js]: I think there are a handful of unused dependencies after removing this block of logic. I've not checked out the branch yet, but at a glance at least `trim` and `config`.

**Comment by aduth** [comment]
What's remaining to be done here that the PR is marked as in progress?

File: client/state/sites/selectors.js (hunk 8 above)
Comment (aduth) [inline client/state/sites/selectors.js]: I agree with the idea of composing these properties from individual selectors. Developers still have the convenience of `getSite` containing computed properties, but can access and maintain those selectors independently (performant, easier to test).

**Comment by aduth** [comment]
FYI planning to introduce new `isSitePreviewable` selector in #7248

**Comment by ehg** [comment]
> What's remaining to be done here that the PR is marked as in progress?

Hopefully nothing. I'm going to mark this as 'awaiting apply' and implement the suggested changes. Should have it ready for final review by EOD, then we can finally get this in.

File: client/lib/site/computed-attributes.js (hunk 12 above)
Comment (ehg) [inline client/lib/site/computed-attributes.js]: Since a fair few tests don't supply `name`, I'd prefer to leave this as it is, to avoid bloating the diff.

File: client/lib/site/index.js (hunk 13 above)
Comment (ehg) [inline client/lib/site/index.js]: I'm leaving `config` in, as it uses `is_previewable`, we'll move it out and use https://github.com/Automattic/wp-calypso/pull/7248 once we've `connect`ed the sidebar.

File: client/state/sites/selectors.js [hunk 14]
```diff
@@ -158,6 +188,26 @@ export function getSiteSlug( state, siteId ) {
...
+ * @param  {Number}  siteId Site ID
+ * @return {?String}        Site domain
+ */
+export function getSiteDomain( state, siteId ) {
+	const site = getRawSite( state, siteId );
+	if ( ! site ) {
+		return null;
+	}
+
+	if ( ( site.options && site.options.is_redirect ) || isSiteConflicting( state, siteId ) ) {
```
Comment (aduth) [inline client/state/sites/selectors.js]: You could reuse `getSiteOption` here:

```js
if ( getSiteOption( state, siteId, 'is_redirect' ) || isSiteConflicting( state, siteId ) ) {
```

At which point you might reorder the function since you're not touching `site` unless these conditions don't apply.

```js
export function getSiteDomain( state, siteId ) {
	if ( getSiteOption( state, siteId, 'is_redirect' ) || isSiteConflicting( state, siteId ) ) {
		return getSiteSlug( state, siteId );
	}
	
	const site = getRawSite( state, siteId );
	if ( ! site ) {
		return null;
	}

	return site.URL.replace( /^https?:\/\//, '' );
}
```

**Comment by ehg** [comment]
@aduth — addressed some of your feedback in https://github.com/Automattic/wp-calypso/pull/6477/commits/eba97e29871464da8d1885512bbb1ca768c14e30 ( thanks! ) — flipping to review.

Also ran the e2e tests on this branch, everything seems fine.

File: client/state/ui/test/selectors.js [hunk 15]
```diff
@@ -30,15 +30,15 @@ describe( 'selectors', () => {
...
+						2916284: { ID: 2916284, name: 'WordPress.com Example Blog', URL: 'https://example.com' }
 					}
 				},
 				ui: {
 					selectedSiteId: 2916284
 				}
 			} );
 
-			expect( selected ).to.eql( { ID: 2916284, name: 'WordPress.com Example Blog' } );
+			expect( selected ).to.contain( { ID: 2916284, name: 'WordPress.com Example Blog' } );
```
Comment (aduth) [inline client/state/ui/test/selectors.js]: Should we just update the asserted comparison object to match the expectation (assuming it adds a few computed properties)?

File: client/state/sites/selectors.js [hunk 16]
```diff
@@ -4,6 +4,7 @@
  * External dependencies
  */
 import {
+	camelCase,
```
Comment (aduth) [inline client/state/sites/selectors.js]: Unused?

File: client/lib/site/index.js (hunk 13 above)
Comment (aduth) [inline client/lib/site/index.js]: Still think this is yet to be addressed https://github.com/Automattic/wp-calypso/pull/6477/files#r73341185

File: client/state/ui/test/selectors.js (hunk 15 above)
Comment (ehg) [inline client/state/ui/test/selectors.js]: :+1:

File: client/state/sites/selectors.js (hunk 16 above)
Comment (ehg) [inline client/state/sites/selectors.js]: :+1:

File: client/lib/site/index.js (hunk 13 above)
Comment (ehg) [inline client/lib/site/index.js]: > Still think this is yet to be addressed https://github.com/Automattic/wp-calypso/pull/6477/files#r73341185

This links to:
> I think there are a handful of unused dependencies after removing this block of logic. I've not checked out the branch yet, but at a glance at least trim and config.

Which I addressed [here](https://github.com/Automattic/wp-calypso/pull/6477/files#r73387334) and [here](https://github.com/Automattic/wp-calypso/pull/6477/files#r69311609) ;)

File: client/state/sites/selectors.js (hunk 14 above)
Comment (ehg) [inline client/state/sites/selectors.js]: Good spot. Fixed.

File: client/lib/site/index.js (hunk 13 above)
Comment (aduth) [inline client/lib/site/index.js]: @ehg : Sorry, I hadn't seen that you'd dropped a comment, but I'm not sure we're on the same page. I'm specifically referring to these newly introduced lint warnings:

```
./node_modules/.bin/eslint --quiet --rule 'no-unused-vars: 2' client/lib/site/index.js 

/Users/andrew/Documents/Code/wp-calypso/client/lib/site/index.js
   9:2  error  'trim' is defined but never used     no-unused-vars
  15:2  error  'config' is defined but never used   no-unused-vars
  19:2  error  'isHttps' is defined but never used  no-unused-vars

✖ 3 problems (3 errors, 0 warnings)
```

Comment (ehg) [inline client/lib/site/index.js]: Ah :facepalm: — I didn't notice you were referring to `lib/site` — long day. Fixed in da7ae9b.

File: client/state/sites/selectors.js [hunk 17]
```diff
@@ -20,17 +20,46 @@ import {
...
+	( state, siteId ) => {
+		const site = getRawSite( state, siteId );
+
+		if ( ! site ) {
+			return null;
+		}
+
+		return {
+			...site,
+			...getComputedAttributes( site ),
```
Comment (aduth) [inline client/state/sites/selectors.js]: My suspicions started tingling that `getComputedAttributes` could have been mutating state here. It turned out not to be true (i.e. everything's fine!) but maybe a case for using `deepFreeze` in selector testing.

**Comment by aduth** [comment]
Could do for some squashing, but this LGTM 👍  Thanks for working on this @ehg

**Comment by ehg** [comment]
Squashed. Thanks for the review all!

**Comment by dmsnell** [comment]
> @ehg : I'd agree with the idea that a normalized site object would be easier to consume from and transition to for developers (evidenced by confusion around and reintroduction of these attributes in the first place 😄 ). Individual selectors have a number of advantages, but "obviousness" (ease of use) isn't really one of them. Composing attributes into the normalized object using individual selectors might be a compromise of these two options, but then there may be confusion around which is preferred (do we have a preference, if the developer has gone through the effort to distinguish them?).

@aduth I'm confused on why we don't want computed attributes here? I think maybe it's the definition of `computed`. For example, if the site object requires normalization, why not do that at the beginning. If it needs to have other objects from within the state tree merged in to feel native (for example, if `site.owner` were to contain the user object of the owner) then that seems like a great use of the selectors. I don't want to have to call a selector to get the site title from `site.site_name` when I should be able to simply access `site.name`

**Comment by aduth** [comment]
@dmsnell : I'm not sure I understand the question. We _do_ want computed attributes. For example, as of this pull request, you can retrieve the `slug` (derived from other site attributes) via either:

```js
slug = getSIte( state, siteId ).slug;
slug = getSiteSlug( state, siteId );
```

The point being made in your quoted text is that it seems to be a bit more obvious for people to work with a single object containing all properties, both raw and computed, than to try to work with a raw site and a number of other independent selectors. But there are trade-offs to this as the single site object will run logic to compute all of the properties, even those not necessarily relevant to the developer's interest. I think we're okay with that, but we optionally also allow for the independent selectors, since (a) they're used in generating the properties anyways, (b) can help avoid cases where we'd otherwise need to use something like [`_.get`](https://lodash.com/docs#get) to avoid calling a property on a `null` (not yet loaded) site, and (c) are easier to test.

I don't think, however, it makes sense to track these properties in state, or at the very least not in persisted state, since it's effectively redundant data (wasted storage).
